Add documentation for Google ads connector (#933)
wso2/docs-mi: en/docs/reference/connectors/google-ads-connector/google-ads-configuration.md

# Google Ads Connector Reference

The following operations allow you to work with the Google Ads API. Click an operation name to see parameter details and samples on how to use it.

## Initialize the connector

To use the Google Ads connector, first create the connection with your configuration. When calling a Google Ads operation, ensure the connection is referenced using the `configKey` attribute.

??? note "init"
    The `init` operation initializes the connection to Google Ads API.
    <table>
        <tr>
            <th>Parameter name</th>
            <th>Description</th>
            <th>Required</th>
        </tr>
        <tr>
            <td><code>name</code></td>

>             <td>Unique name to identify the connection by.</td>

Reviewer: ```suggestion
            <td>Unique name to identify the connection.</td>
```

            <td>Yes</td>
        </tr>
        <tr>
            <td><code>base</code></td>
            <td>The service root URL. The default value is `https://googleads.googleapis.com`.</td>
            <td>Yes</td>
        </tr>
        <tr>
            <td><code>clientId</code></td>
            <td>Client ID of the registered application.</td>
            <td>Yes</td>
        </tr>
        <tr>
            <td><code>clientSecret</code></td>
            <td>Client secret of the registered application.</td>
            <td>Yes</td>
        </tr>
        <tr>
            <td><code>refreshToken</code></td>
            <td>The refresh token that can be used to obtain a new access token.</td>
            <td>Yes</td>
        </tr>
        <tr>
            <td><code>developerToken</code></td>
            <td>The developer token for the Google Ads connection.</td>
            <td>Yes</td>
        </tr>
        <tr>
            <td><code>loginCustomerId</code></td>

>             <td>The ID of the manager account. Required when using a manager account to access an customer account.</td>

Reviewer: ```suggestion
            <td>The ID of the manager account. Required when using a manager account to access a customer account.</td>
```

            <td>No</td>
        </tr>
        <tr>
            <td><code>tokenEndpoint</code></td>

>             <td>The HTTP endpoint that can be uses to obtain an access token. The default value is `https://www.googleapis.com/oauth2/v3/token`.</td>

Reviewer: ```suggestion
            <td>The HTTP endpoint that can be used to obtain an access token. The default value is `https://www.googleapis.com/oauth2/v3/token`.</td>
```

            <td>Yes</td>
        </tr>
    </table>

    **Sample configuration**

    ```xml
    <googleAds.init>
        <connectionType>googleAds</connectionType>

>         <name>GGOGLE_ADS_CONN</name>

Reviewer: Is `GGOGLE` a typo in the documentation? There are multiple occurrences.

Author: Fixed with [70b7f97](https://github.com/wso2/docs-mi/pull/933/commits/70b7f970e86199fba8829404016b58ac512ab3a9)

        <base>https://googleads.googleapis.com</base>
        <clientId>REPLACE_WITH_CLIENT_ID</clientId>
        <clientSecret>REPLACE_WITH_CLIENT_SECRET</clientSecret>
        <refreshToken>REPLACE_WITH_REFRESH_TOKEN</refreshToken>
        <developerToken>REPLACE_WITH_DEVELOPER_TOKEN</developerToken>
        <loginCustomerId>REPLACE_WITH_MANAGER_ID</loginCustomerId>
        <tokenEndpoint>https://www.googleapis.com/oauth2/v3/token</tokenEndpoint>
    </googleAds.init>
    ```

## Operations

??? note "createCustomerClient"
    The `createCustomerClient` operation creates a new client under the provided manager. The new client customer is returned.
    <table>
        <tr>
            <th>Parameter name</th>
            <th>Description</th>
            <th>Required</th>
        </tr>
        <tr>
            <td><code>customerId</code></td>

>             <td>The ID of the manager under whom client customer is being created. Type: string</td>

Reviewer: Shall we format the param types as well?

`Type: <code>string</code>`

There are multiple occurrences.

Author: Fixed with [70b7f97](https://github.com/wso2/docs-mi/pull/933/commits/70b7f970e86199fba8829404016b58ac512ab3a9)

>             <td>The ID of the manager under whom client customer is being created. Type: string</td>

Reviewer: ```suggestion
            <td>The ID of the manager under whom the client customer is being created. Type: string</td>
```

            <td>Yes</td>
        </tr>
        <tr>
            <td><code>customerClient</code></td>

>             <td>The new client customer to create. The resource name on this customer will be ignored. Type: object [Customer](https://developers.google.com/google-ads/api/rest/reference/rest/v17/Customer)</td>

Reviewer: Shall we update the external links as below:

`For more information, go to [Customer](https://developers.google.com/google-ads/api/rest/reference/rest/v17/Customer).`

There are multiple occurrences.

Author: Fixed with [3d8b309](https://github.com/wso2/docs-mi/pull/933/commits/3d8b309f07637aea9f374fc5b18c0cf72322444f)

            <td>Yes</td>
        </tr>
        <tr>
            <td><code>accessRole</code></td>

>             <td>The proposed role of user on the created client customer. Accessible only to customers on the allow-list. Type: enum [AccessRole](https://developers.google.com/google-ads/api/rest/reference/rest/v17/AccessRole)</td>

Reviewer: ```suggestion
            <td>The proposed role of the user on the created client customer. Accessible only to customers on the allow-list. Type: enum [AccessRole](https://developers.google.com/google-ads/api/rest/reference/rest/v17/AccessRole)</td>
```

        "response_type": "MUTABLE_RESOURCE",
        "operation": {
            "update": {
                "resourceName": "customers/456456456",
                "descriptiveName": "My Client 1",
                "hasPartnersBadge": true
            },
            "updateMask": "descriptiveName,hasPartnersBadge"
        }
    }
    ```

??? note "campaignBudgets"
    The `campaignBudgets` operation creates, updates, or removes campaign budgets. Operation statuses are returned.
    <table>
        <tr>
            <th>Parameter name</th>
            <th>Description</th>
            <th>Required</th>
        </tr>
        <tr>
            <td><code>customerId</code></td>
            <td>The ID of the customer whose campaign budgets are being modified. Type: string</td>
            <td>Yes</td>
        </tr>
        <tr>
            <td><code>operations</code></td>
            <td>The list of operations to perform on individual campaign budgets. Type: object [CampaignBudgetOperation](https://developers.google.com/google-ads/api/rest/reference/rest/v17/CampaignBudgetOperation)</td>
            <td>Yes</td>
        </tr>
        <tr>
            <td><code>partialFailure</code></td>

>             <td>If true, successful operations will be carried out and invalid operations will return errors. If false, all operations will be carried out in one transaction if and only if they are all valid. Default is false. Type: boolean</td>

Author: Fixed with [13cd302](https://github.com/wso2/docs-mi/pull/933/commits/13cd302a06eb72f432461d655f7426a4238075cb)

                    "updateMask": "amountMicros"
                }
            ]
        }
        ```
    === "Remove"
        ```json
        {
            "customer_id": "456456456",
            "operations": [
                {
                    "remove": "customers/456456456/campaignBudgets/789789789"
                }
            ]
        }
        ```

??? note "campaignsMutate"
    The `campaignsMutate` operation creates, updates, or removes campaigns. Operation statuses are returned.
    <table>
        <tr>
            <th>Parameter name</th>
            <th>Description</th>
            <th>Required</th>
        </tr>
        <tr>
            <td><code>customerId</code></td>
            <td>The ID of the customer whose campaigns are being modified. Type: string</td>
            <td>Yes</td>
        </tr>
        <tr>
            <td><code>operations</code></td>
            <td>The list of operations to perform on individual campaigns. Type: object [CampaignOperation](https://developers.google.com/google-ads/api/rest/reference/rest/v17/CampaignOperation)</td>
            <td>Yes</td>
        </tr>
        <tr>
            <td><code>partialFailure</code></td>

>             <td>If true, successful operations will be carried out and invalid operations will return errors. If false, all operations will be carried out in one transaction if and only if they are all valid. Default is false. Type: boolean</td>

Reviewer: ```suggestion
            <td>If true, successful operations will be carried out and invalid operations will return errors. If false, all operations will be carried out in one transaction if and only if they are all valid. The default value is <code>false</code>. Type: boolean</td>
```

                    }
                }
            ]
        }
        ```
    === "Remove"
        ```json
        {
            "customer_id": "1231231223",
            "validateOnly": false,
            "operations": [
                {
                    "remove": "customers/1231231223/campaigns/789789789"
                }
            ]
        }
        ```

??? note "adGroupsMutate"
    The `adGroupsMutate` operation creates, updates, or removes ad groups. Operation statuses are returned.
    <table>
        <tr>
            <th>Parameter name</th>
            <th>Description</th>
            <th>Required</th>
        </tr>
        <tr>
            <td><code>customerId</code></td>
            <td>The ID of the customer whose ad groups are being modified. Type: string</td>
            <td>Yes</td>
        </tr>
        <tr>
            <td><code>operations</code></td>
            <td>The list of operations to perform on individual ad groups. Type: object [AdGroupOperation](https://developers.google.com/google-ads/api/rest/reference/rest/v17/AdGroupOperation)</td>
            <td>Yes</td>
        </tr>
        <tr>
            <td><code>partialFailure</code></td>

>             <td>If true, successful operations will be carried out and invalid operations will return errors. If false, all operations will be carried out in one transaction if and only if they are all valid. Default is false. Type: boolean</td>

Reviewer: ```suggestion
            <td>If true, successful operations will be carried out and invalid operations will return errors. If false, all operations will be carried out in one transaction if and only if they are all valid. The default value is <code>false</code>. Type: boolean</td>
```

Author: Fixed with [13cd302](https://github.com/wso2/docs-mi/pull/933/commits/13cd302a06eb72f432461d655f7426a4238075cb)

                        "cpmBidMicros": "10000"
                    }
                }
            ]
        }
        ```
    === "Remove"
        ```json
        {
            "customer_id": "123123123",
            "operations": [
                {
                    "remove": "customers/123123123/adGroups/456456456"
                }
            ]
        }
        ```

??? note "adGroupAdsMutate"
    The `adGroupAdsMutate` operation creates, updates, or removes ads. Operation statuses are returned.
    <table>
        <tr>
            <th>Parameter name</th>
            <th>Description</th>
            <th>Required</th>
        </tr>
        <tr>
            <td><code>customerId</code></td>
            <td>The ID of the customer whose ads are being modified. Type: string</td>
            <td>Yes</td>
        </tr>
        <tr>
            <td><code>operations</code></td>
            <td>The list of operations to perform on individual ads. Type: object [AdGroupAdOperation](https://developers.google.com/google-ads/api/rest/reference/rest/v17/AdGroupAdOperation)</td>
            <td>Yes</td>
        </tr>
        <tr>
            <td><code>partialFailure</code></td>

>             <td>If true, successful operations will be carried out and invalid operations will return errors. If false, all operations will be carried out in one transaction if and only if they are all valid. Default is false. Type: boolean</td>

Reviewer: ```suggestion
            <td>If true, successful operations will be carried out and invalid operations will return errors. If false, all operations will be carried out in one transaction if and only if they are all valid. The default value is <code>false</code>. Type: boolean</td>
```

Author: Fixed with [13cd302](https://github.com/wso2/docs-mi/pull/933/commits/13cd302a06eb72f432461d655f7426a4238075cb)

                    }
                }
            ]
        }
        ```
    === "Remove"
        ```json
        {
            "customer_id": "123123123",
            "operations": [
                {
                    "remove": "customers/123123123/adGroupAds/456456456~789789789"
                }
            ]
        }
        ```

??? note "adsMutate"
    The `adsMutate` operation updates ads. Operation statuses are returned. Updating ads is not supported for TextAd, ExpandedDynamicSearchAd, GmailAd and ImageAd.
    <table>
        <tr>
            <th>Parameter name</th>
            <th>Description</th>
            <th>Required</th>
        </tr>
        <tr>
            <td><code>customerId</code></td>
            <td>The ID of the customer whose ads are being modified. Type: string</td>
            <td>Yes</td>
        </tr>
        <tr>
            <td><code>operations</code></td>
            <td>The list of operations to perform on individual ads. Type: object [AdOperation](https://developers.google.com/google-ads/api/rest/reference/rest/v17/AdOperation)</td>
            <td>Yes</td>
        </tr>
        <tr>
            <td><code>partialFailure</code></td>

>             <td>If true, successful operations will be carried out and invalid operations will return errors. If false, all operations will be carried out in one transaction if and only if they are all valid. Default is false. Type: boolean</td>

Reviewer: ```suggestion
            <td>If true, successful operations will be carried out and invalid operations will return errors. If false, all operations will be carried out in one transaction if and only if they are all valid. The default value is <code>false</code>. Type: boolean</td>
```

Author: Fixed with [13cd302](https://github.com/wso2/docs-mi/pull/933/commits/13cd302a06eb72f432461d655f7426a4238075cb)

        <customerId>{json-eval($.customer_id)}</customerId>
        <operations>{json-eval($.operations)}</operations>
    </googleAds.adsMutate>
    ```
 
    **Sample request**

    ```json
    {
        "customer_id": "123123123",
        "operations": [
            {
                "updateMask": "finalUrls",
                "update": {
                    "resourceName": "customers/123123123/ads/789789789",
                    "finalUrls": [
                        "https://www.github.com"
                    ]
                }
            }
        ]
    }
    ```

??? note "search"
    The `search` operation returns all rows that match the search query.
    <table>
        <tr>
            <th>Parameter name</th>
            <th>Description</th>
            <th>Required</th>
        </tr>
        <tr>
            <td><code>customerId</code></td>
            <td>The ID of the customer being queried. Type: string</td>
            <td>Yes</td>
        </tr>
        <tr>
            <td><code>query</code></td>
            <td>The query string. Type: string</td>
            <td>Yes</td>
        </tr>
        <tr>
            <td><code>pageToken</code></td>

>             <td>Token of the page to retrieve. If not specified, the first page of results will be returned. Use the value obtained fromnextPageTokenin the previous response in order to request the next page of results. Type: string</td>

Reviewer: ```suggestion
            <td>Token of the page to retrieve. If not specified, the first page of the results will be returned. Use the value obtained fromnextPageTokenin the previous response to request the next page of results. Type: string</td>
```

Author: Fixed with [70b7f97](https://github.com/wso2/docs-mi/pull/933/commits/70b7f970e86199fba8829404016b58ac512ab3a9)

>             <td>Token of the page to retrieve. If not specified, the first page of results will be returned. Use the value obtained fromnextPageTokenin the previous response in order to request the next page of results. Type: string</td>

Reviewer: Please fix the words `fromnextPageTokenin`.

Author: Fixed with [70b7f97](https://github.com/wso2/docs-mi/pull/933/commits/70b7f970e86199fba8829404016b58ac512ab3a9)

            <td>No</td>
        </tr>
        <tr>
            <td><code>pageSize</code></td>
            <td>Number of elements to retrieve in a single page. When too large a page is requested, the server may decide to further limit the number of returned resources. Type: integer</td>
            <td>No</td>
        </tr>
        <tr>
            <td><code>validateOnly</code></td>
            <td>If true, the request is validated but not executed. Type: boolean</td>
            <td>No</td>
        </tr>
        <tr>
            <td><code>returnTotalResultsCount</code></td>
            <td>If true, the total number of results that match the query ignoring the LIMIT clause will be included in the response. Default is false. Type: boolean</td>
            <td>No</td>
        </tr>
        <tr>
            <td><code>summaryRowSetting</code></td>

>             <td>Determines whether a summary row will be returned. By default, summary row is not returned. If requested, the summary row will be sent in a response by itself after all other query results are returned. Type: enum [SummaryRowSetting](https://developers.google.com/google-ads/api/rest/reference/rest/v17/SummaryRowSetting)</td>

Reviewer: ```suggestion
            <td>Determines whether a summary row will be returned. By default, a summary row is not returned. If requested, the summary row will be sent in a response by itself after all other query results are returned. Type: enum [SummaryRowSetting](https://developers.google.com/google-ads/api/rest/reference/rest/v17/SummaryRowSetting)</td>
```

            <td>No</td>
        </tr>
    </table>

    **Sample configuration**

    ```xml
    <googleAds.search configKey="GGOGLE_ADS_CONN">
        <customerId>{json-eval($.customer_id)}</customerId>
        <query>{json-eval($.query)}</query>
    </googleAds.search>
    ```
 
    **Sample request**

    ```json
    {
        "customer_id": "123123123",
        "query": "SELECT customer_client.client_customer FROM customer_client"
    }
    ```

## Error codes related to Google Ads Connector

> The connector may encounter errors during operation execution. When an error occurs, the `ERROR_MESSAGE` property will contain detailed information about the error. You can handle these errors using a `Fault Sequence` in your integration. For more information, refer to [Using Fault Sequences]({{base_path}}/learn/examples/sequence-examples/using-fault-sequences/).

Reviewer: ```suggestion
The connector may encounter errors during operation execution. When an error occurs, the `ERROR_MESSAGE` property will contain detailed information about the error. You can handle these errors using a `Fault Sequence` in your integration. For more information, see [Using Fault Sequences]({{base_path}}/learn/examples/sequence-examples/using-fault-sequences/).
```

| Error code | Description |
| -------- | ------- |

> | 701001 | A general error has occured. |

Reviewer: ```suggestion
| 701001 | A general error has occurred. |
```

> | 701002 | An error has occured due to an invalid configuration. |

Reviewer: ```suggestion
| 701002 | An error has occurred due to an invalid configuration. |
```

> | 701003 | An error has occured in the access token generation flow. |

Reviewer: ```suggestion
| 701003 | An error has occurred in the access token generation flow. |
```

> | 701003 | An error has occurred in the access token generation flow. |

Reviewer: Shall we reword the description of each error code as follows (for clarity): 
- General error
- Invalid configuration error
- Error in access token generation flow
